### PR TITLE
Datumizes saymodes

### PR DESCRIPTION
**Pull request by ghost**
Swarmer chat, blob chat, alien chat, ling chat, silicon chat, drone chat, golem chat, ALL DATUMIZED!

### CODE REVIEW

**Comment by MoreRobustThanYou** [comment]
Yes, I tested it. it works.

**Comment by deathride58** [comment]
Doin god's work

File: code/modules/mob/living/say.dm [hunk 1]
```diff
@@ -134,8 +134,12 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	// AIs use inherent channels for the holopad. Most inherent channels
 	// ignore the language argument however.
 
-	if(handle_inherent_channels(message, message_mode, language)) //Hiveminds, binary chat & holopad.
-		return
+	var/key = lowertext(copytext(message, 2, 3))
+	for(var/_SM in subtypesof(/datum/saymode))
+		var/datum/saymode/SM = new _SM()
+		if((SM.key == key || SM.mode == message_mode) && SM.handle_message(src, message, language))
```
Comment (ninjanomnom) [inline code/modules/mob/living/say.dm]: \>Every time anyone says anything loop through every saymode type

**Review by ninjanomnom — DISMISSED** [review]

File: code/modules/mob/living/say.dm (hunk 1 above)
Comment (RemieRichards) [reply code/modules/mob/living/say.dm]: worst part: NOT CACHING THE STATELESS DATUMS.

Comment (MoreRobustThanYou) [reply code/modules/mob/living/say.dm]: @RemieRichards what's a stateless datum?

Comment (RemieRichards) [reply code/modules/mob/living/say.dm]: a datum that stores no state.
it comes from just the general concept of a stateless object in programming.

these saymodes always pass in their user/anything they need to use as arguments to procs, they don't ever store anything on themselves, meaning it's perfectly valid (and is in fact the BEST WAY TO DO THIS) to keep only 1 copy of each, shared between everything that speaks.

File: code/modules/mob/living/living.dm [hunk 2]
```diff
@@ -1010,3 +1010,14 @@
 /mob/living/proc/add_abilities_to_panel()
 	for(var/obj/effect/proc_holder/A in abilities)
 		statpanel("[A.panel]",A.get_panel_text(),A)
+
+/mob/living/lingcheck() //1 is ling w/ no hivemind. 2 is ling w/hivemind. 3 is ling victim being linked into hivemind.
+	if(mind)
+		var/datum/antagonist/changeling/changeling = mind.has_antag_datum(/datum/antagonist/changeling)
+		if(changeling)
+			if(changeling.changeling_speak)
+				return 2
```
Comment (kevinz000) [inline code/modules/mob/living/living.dm]: m-magical n-numbers...

File: code/modules/mob/living/say.dm (hunk 1 above)
Comment (kevinz000) [reply code/modules/mob/living/say.dm]: @MoreRobustThanYou do saymodes ever get their vars changed if not why don't you just make a global list of them or something that's associative?

Comment (kevinz000) [reply code/modules/mob/living/say.dm]: and after i typed that i realized it's basically what remie said :P

File: code/modules/mob/living/living.dm (hunk 2 above)
Comment (MoreRobustThanYou) [reply code/modules/mob/living/living.dm]: this code already existed lmao

File: code/modules/mob/living/say.dm [hunk 3]
```diff
@@ -134,8 +135,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	// AIs use inherent channels for the holopad. Most inherent channels
 	// ignore the language argument however.
 
-	if(handle_inherent_channels(message, message_mode, language)) //Hiveminds, binary chat & holopad.
-		return
+	if(key in GLOB.saymode_keys)
+		for(var/datum/saymode/SM in GLOB.saymodes)
```
Comment (optimumtact) [inline code/modules/mob/living/say.dm]: just make this a list keyed by the key
`[key] => saymode datum`

File: code/controllers/subsystem/radio.dm [hunk 4]
```diff
@@ -1,9 +1,15 @@
 SUBSYSTEM_DEF(radio)
 	name = "Radio"
-	flags = SS_NO_FIRE|SS_NO_INIT
+	flags = SS_NO_FIRE
```
Comment (Cyberboss) [inline code/controllers/subsystem/radio.dm]: Revert this

File: code/controllers/subsystem/radio.dm [hunk 5]
```diff
@@ -1,9 +1,15 @@
...
 
 	var/list/datum/radio_frequency/frequencies = list()
 
+/datum/controller/subsystem/radio/Initialize(timeofday)
```
Comment (Cyberboss) [inline code/controllers/subsystem/radio.dm]: Use PreInit

File: code/datums/saymode.dm [hunk 6]
```diff
@@ -0,0 +1,112 @@
+GLOBAL_LIST_EMPTY(saymodes)
```
Comment (Cyberboss) [inline code/datums/saymode.dm]: Should be in radio SS

File: code/datums/saymode.dm [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+/datum/saymode
+	var/key
+	var/mode
+
+/datum/saymode/proc/handle_message(mob/living/user, message, datum/language/language)
+	return FALSE
```
Comment (Cyberboss) [inline code/datums/saymode.dm]: I'd invert the logic here, so things only return TRUE if they weren't handled. Then you could eliminate all the return TRUEs in the overrides

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by SpaceManiac** [comment]
`changeling/handle_message` is still using magic numbers in a couple places.

Also I would check to make sure that a message like "Type" doesn't trigger detecting the `y` mode.

**Comment by MoreRobustThanYou** [comment]
I ensured that saying things like "AGGGGGGGGGGGGGGGGGGGGGGGGG!" won't trigger changeling chat

File: code/datums/saymode.dm [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+				if(M in GLOB.dead_mob_list)
+					var/link = FOLLOW_LINK(M, user)
+					to_chat(M, "[link] [msg]")
+				else
+					switch(M.lingcheck())
+						if(3)
+							to_chat(M, msg)
+						if(2)
+							to_chat(M, msg)
+						if(1)
```
Comment (coiax) [inline code/datums/saymode.dm]: These need to be the defines.

**Comment by Cyberboss** [comment]
Do what coiax said

**Comment by MoreRobustThanYou** [comment]
@Cyberboss i did

**Comment by Cyberboss** [comment]
Fix your shit

File: code/datums/saymode.dm [hunk 9]
```diff
@@ -0,0 +1,108 @@
+/datum/saymode
+	var/key
+	var/mode
+
+/datum/saymode/proc/handle_message(mob/living/user, message, datum/language/language)
```
Comment (Cyberboss) [inline code/datums/saymode.dm]: Document what the vars and proc return value does

File: code/modules/mob/living/say.dm [hunk 10]
```diff
@@ -134,8 +135,11 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	// AIs use inherent channels for the holopad. Most inherent channels
 	// ignore the language argument however.
 
-	if(handle_inherent_channels(message, message_mode, language)) //Hiveminds, binary chat & holopad.
-		return
+	if(key && SSradio.saymodes[key])
+		var/datum/saymode/SM = SSradio.saymodes[key]
```
Comment (Cyberboss) [inline code/modules/mob/living/say.dm]: Don't lookup the value twice

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**